typify-impl/src/enums.rs

```diff
@@ -609,28 +609,47 @@ impl TypeSpace {
             // account for types such as Uuid whose names come from outside of
             // the schema... but you can't win them all.
             .map(schema_is_named)
-            .collect::<Option<Vec<_>>>()
-            // Prune the common prefixes from all variant names. If this
-            // results in any of them being empty, we don't use these names.
-            .and_then(|variant_names| {
-                let common_prefix = variant_names
-                    .iter()
-                    .cloned()
-                    .reduce(|a, b| get_common_prefix(&a, &b))
-                    .unwrap();
-                variant_names
-                    .into_iter()
-                    .map(|var_name| {
-                        let var_name = &var_name[common_prefix.len()..];
-                        if var_name.is_empty() {
-                            None
-                        } else {
-                            Some(var_name.to_string())
-                        }
-                    })
-                    .collect::<Option<Vec<_>>>()
+            .collect::<Vec<_>>();
+
+        let common_prefix = variant_names
+            .iter()
+            .flatten()
+            .cloned()
+            .reduce(|a, b| get_common_prefix(&a, &b))
+            .unwrap_or_else(String::new);
+
+        // Store seen, un-stripped names here.
+        let duplicate_names = variant_names
+            .iter()
+            .flatten()
+            .cloned()
+            .fold(HashMap::new(), |mut counter: HashMap<_, u8>, name| {
+                *counter.entry(name).or_default() += 1;
+                counter
             })
-            // Fall back to `VariantN` naming.
+            .into_iter()
+            .filter_map(|(name, count)| if count > 1 { Some(name) } else { None })
+            .collect::<HashSet<_>>();
+
+        // Prune the common prefixes from all variant names. If this
+        // results in any of them being empty, we don't use these names.
+        //
+        // Fall back to VariantN if ALL variant names cannot be derived from
+        // the schema.
+        let variant_names = variant_names
+            .into_iter()
+            .map(|var_name| {
+                match var_name {
+                    None => None,
+                    var_name => {
+                        var_name.filter(|var_name| !duplicate_names.contains(var_name.as_str()))
+                    }
+                }
+                .map(|var_name| var_name[common_prefix.len()..].to_string())
+                .filter(|var_name| !var_name.is_empty())
+                .map(|var_name| var_name.to_string())
```

Comment:
is this necessary? it looks like two lines above we've already done the `to_string()`